audio/out/buffer.c

```diff
@@ -425,10 +425,21 @@ void ao_set_paused(struct ao *ao, bool paused, bool eof)
 
     mp_mutex_unlock(&p->lock);
 
-    if (do_reset)
-        ao->driver->reset(ao);
-    if (do_start)
-        ao->driver->start(ao);
+    if (ao->driver->set_pause) {
+        if (do_reset) {
```

Comment:
For pull-based ao, these two variables are set in https://github.com/mpv-player/mpv/blob/44489567fc4486a45f1d02a7d0ed5c273ff35230/audio/out/buffer.c#L405-L409 and https://github.com/mpv-player/mpv/blob/44489567fc4486a45f1d02a7d0ed5c273ff35230/audio/out/buffer.c#L418-L420

The reason why we need to put code invoking `set_pause` here is to avoid deadlock. `p->lock` should be unlocked before `set_pause` is called, otherwise the ao may get deadlocked when acquiring `p->lock`.

Reply:
I have not experienced any locking issue, but if that is the case it would still be clearer to add a new `do_pause`.

Reply:
> I have not experienced any locking issue

Because currently no ao will acquire `p->lock` (e.g., by using `ao_read_data`) during the call to `set_pause`. But this behavior is not forbidden by API design, and ao have the ability to do that.

> it would still be clearer to add a new do_pause.

Ok